[DRAFT]

WesleyTangNationalInstruments opened the pull request:
- [ ] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

TODO: Check the above box with an 'x' indicating you've read and followed [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

TODO: Include high-level description of the changes in this pull request.

### Why should this Pull Request be merged?

TODO: Justify why this contribution should be part of the project.

### What testing has been done?

TODO: Detail what testing has been done to ensure this submission meets requirements.

pbirkhol-ni commented on Source/Runtime/MeasurementLink Measurement Server/Helpers/Get Results Metadata.vi:
- There are still quite a few reference that don't get closed: The Panel reference and both arrays of Controls references.
- This should be split up into logical subVIs. Those subVIs can probably be used in both this VI and in `Get Configuration Metadata.vi.`
- There is no need to append `output` to the Parameter name (and doing so will cause issues). Because of where we are calling this VI, we already know they are outputs and they are going to get bundled into the `outputs` element of the `measurement_signature` cluster.
- The logic for handling enum arrays is incorrect. We want the refnum of the array, not a refnum of an array element. So the logic should be something like: `if enum or if enum array then cache control refnum`. The cached control will always be either an enum control refnum or an enum array control refnum. This means that you might have to modify your `Create Enum Type Specialization.vi` to work with both an enum refnum and an enum array refnum. 
- You can add a conditional terminal to a for loop by right clicking on it and selecting `Conditional Terminal`. I think that would make your logic simpler when determining if the enum specialization already exists (you can stop the for loop early if/when you find a match). 
- I don't think using `Insert Into Array` is the correct thing to do when adding an enum type specialization. You want `Build Array` instead. Also, anything you insert into the array is going to get lost since you don't save it in a shift register. But saving it in a shift register is weird because the inner for loop iterates over the type specialization array and you wouldn't want it to iterate over new elements (although it probably wouldn't hurt). So I guess you could just use a shift register. You could also make a copy of the array outside of the outer for loop so that one copy stays the same and the other copy is the shift register copy

WesleyTangNationalInstruments replied:
I addressed these comments as suggested in my latest update.

WesleyTangNationalInstruments replied:
> This should be split up into logical subVIs. 

I made subVIs for the getting of the controls and the filtering to enum controls, but I left the logic where we translate Type Specializations into Metadata as that seems to be the core function of "getting metadata". If you would prefer I move that logic into a subVI as well, let me know.

pbirkhol-ni commented on Source/Runtime/MeasurementLink Measurement Server/Helpers/Get Configuration Metadata.vi:
This VI has similar feedback to `Get Results Metadata.vi`. In addition:
- This could _probably_ be merged with `Get Results Metadata.vi`, but I think I'd rather keep them separate and then refactor the internal logic into subVIs that get reused. That way we're not duplicating a lot of code but we also have more flexibility for the two implementations to diverge in the future (if needed). I don't feel strongly about this, though. My concern is that it will be more difficult to understand the logic if we try to have one VI called `Get Metadata` that handles both inputs and outputs.

WesleyTangNationalInstruments replied:
I left these two VIs (Get Configuration MetaData.vi and Get Results Metadata.vi) as separate VIs as suggested, though they are 100% logically identical after suggested refactors.

WesleyTangNationalInstruments commented on Source/Runtime/MeasurementLink Measurement Server/Classes/MeasurementPluginService/Get Type Specializations.vi:
Yes, this VI refnum was linking to the old version which did not contain the added enum values for Type Specialization and Type Specialization Key.

![image](https://github.com/ni/measurementlink-labview/assets/103216564/35f93adc-efd3-4976-8568-9211bb7c582d)

WesleyTangNationalInstruments commented:
Create Enum Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/fbb89391-033f-4831-a776-348e20efc0ec)

Filter for Enum Controls.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/aac8b40c-48f9-4d9f-b765-f852aca715ec)

Get Measurement Configuration Controls.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/20dcc8ef-cbad-432a-8e14-350362d7a758)

Get Configuration MetaData.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/72e07985-c1fd-40fe-b2d0-e7558411864c)

Get Results MetaData.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/7e9421c8-3aed-4926-9c99-3bdd0a2cbbf1)

Get MetaData from Configurations.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/e0aa03a0-4afa-4e53-8691-b66837e89364)

Get MetaData from Results.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/421a2222-adc8-4533-8f92-cf17c6d3a2c2)

Get Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/c3d23823-746e-49bf-bc3d-e3242bfa8d0d)

jasonmreding commented:
Create Enum Type Specialization:

- We are always adding enum.values annotation even if a control isn't an enum. If it isn't an enum, the list of enum values is empty so having the annotation at all is confusing. I would prefer not to generate any enum annotations if the control isn't an enum.
- Actually.... it looks like you are only calling this VI after you have already filtered the list of controls to enums. In that case, you should update the control ref in/out to an enum control reference and delete all of the code that validates the control is an enum.
- Rather than require the caller to pass in the name of the parameter by extracting the text for the label of the control, just move it into the implementation of the VI.
- You shouldn't need the upcasts to generic control class. If you want to leave them to get rid of the coercion dot (type widening) due to the implicit upcast by LV for the wire tunnel, I won't complain. However, I don't generally bother with explicit upcasts just to get rid of the coercion dot. This comment likely applies throughout.
- Consider creating IsEnumControl, IsArrayControl, or IsEnumOrEnumArrayControl helper VIs for readability.
- Make this helper private like the rest of them.

Get Results Metadata:

- This is using the control for the configuration rather than the results. If we haven't already, we should test enum indicators work as expected.
- Consolidate logic between this VI and Get Configuration Metadata. Either pass in the control reference for the metadata or create a shared sub VI for the common parts.

WesleyTangNationalInstruments commented:
> Actually.... it looks like you are only calling this VI after you have already filtered the list of controls to enums. In that case, you should update the control ref in/out to an enum control reference and delete all of the code that validates the control is an enum.

I don't think we can use an enum control reference here as we pass in both controls references of both enum and arrays of enums as valid input.

> This is using the control for the configuration rather than the results. 

Are you referring to `ConfigurationParameter_AnnotationsEntry.ctl` being used in both `Get Configuration Metadata.vi` and `Get Results Metadata.vi`? I think this is really just a naming problem as the AnnotationsEntry gets used by both `ConfigurationParameter.ctl` and `Output.ctl` now and should be generically named. I don't believe we are incorrectly using `ConfigurationParameter.ctl` as that is handled at the `Get MetaData from Results.vi` level rather than here in the `Get Results Metadata.vi` which is agnostic of that type.

> If we haven't already, we should test enum indicators work as expected.

I don't have a fully working example that behaviors correctly in InstrumentStudio yet since there's an issue with serialization/deserialization (which I believe needs to be resolved outside of this LabView code), but I have tested a service with a debugger to verify that an example with an enum will run through this code and produce annotations. 

Screenshots below:

Create Enum Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/d250766f-e4c6-479c-bf5b-3a4ec5008c14)

Filter for Enum Controls.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/ad251607-f437-4d35-9761-17110a2c172c)

Get Configuration Metadata.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/59bc79af-797c-4035-9ac6-7c774b70674e)

Get Enum Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/29fb58fc-891a-44db-b179-d7f8bfcb53e4)

Get Measurement Configuration Controls.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/fa36dc58-2f25-429d-8ecc-b6f8062e92b4)

Get MetaData from Configurations.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/d9b73f28-c5b5-4db1-8dd6-e9032e0341ed)

Get MetaData from Results.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/7d7ef33e-34c1-436e-8bfc-53679b24d12e)

Get Metadata.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/f186598d-adbe-44df-a373-bfb134e66851)

Get Results Metadata.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/83690a23-b7f8-437c-a6d4-bc8f1a39eb93)

IsEnumArrayControl.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/811bb113-5208-46fe-8936-d95c8ab9390c)

IsEnumControl.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/1e66c0be-40f6-4c2b-a63f-9eaee72bc56a)

IsEnumOrEnumArrayControl.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/6ff36988-8847-4fd8-b617-8d79b559ac00)

jasonmreding commented:
> I don't think we can use an enum control reference here as we pass in both controls references of both enum and arrays of enums as valid input.

The array control also has a control reference for the array element, which we get when generating the enum specialization. What I was suggesting is that the Filter for Enum Controls.vi extract this control reference for array controls (casted as enum control reference) and the original control reference for enum controls (casted as enum reference). I think if the various helpers for Is Enum, Is Enum Array, and Is Enum or Enum Array had an output for the casted reference type this would be a little easier and clean up other parts of the diagram throughout. The control ref out could be changed to the more specific type, but I think this might be more confusing than just having a second ref out for the more specific type of control reference. I still like having the Boolean output since it saves the caller from having to check/call Not A Refnum.

> Are you referring to ConfigurationParameter_AnnotationsEntry.ctl being used in both Get Configuration Metadata.vi and Get Results Metadata.vi? I think this is really just a naming problem as the AnnotationsEntry gets used by both ConfigurationParameter.ctl and Output.ctl now and should be generically named. I don't believe we are incorrectly using ConfigurationParameter.ctl as that is handled at the Get MetaData from Results.vi level rather than here in the Get Results Metadata.vi which is agnostic of that type.

I'm referring to this which is called from both Get Configuration Metadata and Get Results Metadata. If I'm reading this correctly, that means we're going to generate enum annotations based on the configuration controls in both cases.

![image](https://github.com/ni/measurementlink-labview/assets/103200169/7dada755-c0ff-4b6e-99a4-e228c0c74077)

- Get MetaData from Results and Get MetaData from Configuration appear to be entirely duplicated minus the sub VI call to get the metadata that is specific to either the configuration or results. I would recommend refactoring so that either the annotations are passed into both or tunnel through the VI refnum for the appropriate ctl so the sub VI is no longer specific to configuration or results.
- IsEnumArrayControl should use IsEnumControl in its implementation.

WesleyTangNationalInstruments commented:
> Get MetaData from Results and Get MetaData from Configuration appear to be entirely duplicated minus the sub VI call to get the metadata that is specific to either the configuration or results. I would recommend refactoring so that either the annotations are passed into both or tunnel through the VI refnum for the appropriate ctl so the sub VI is no longer specific to configuration or results.

@jasonmreding I addressed this by refactoring the `Get MetaData from Results.vi` and `Get MetaData from Configuration.vi` so that we call into the non-specific `Get Metadata.vi` and allows us to remove both the specific `Get Configuration Metadata.vi` and `Get Results Metadata.vi` as well as creating a `Get Annotations For Paramater.vi`. 

I addressed the other comments as suggested.

Also note the additional change to `LabVIEW datatype to Measurement datatype.vi` so that instead of coercing LabVIEW enums to UInts, we use the gRPC TYPE_ENUM. This was something Everett and I ran into while trying to get a fully functional example with these changes. 

IsEnumArrayControl.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/9c5324c9-331d-443d-ad80-59a9e720dd9a)

IsEnumControl.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/2f4a40b9-102f-462c-a436-a10aa85d1b4b)

Filter for Enum Controls.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/e980ef15-c436-454a-9030-37a34c356305)

Create Enum Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/ec504799-6bd3-40e1-bdad-d33418782e33)

Get Enum Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/583e68f7-7878-4d01-b84a-c3f2bef3b150)

Get MetaData from Configuration.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/d8b618e6-d6a5-4643-991d-0cdc32e50d1d)

Get MetaData from Results.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/cd756d22-98a5-4379-b939-de9ae84f12d5)

Get Metadata.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/1b287c08-e9e3-42ff-bf0f-fdd9185b2cc2)

Get Annotations For Parameter.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/907780ff-f246-42fd-ba93-11cddfc63c24)

LabVIEW datatype to Measurement datatype.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/0b451740-08e5-4f2f-9cf6-d9c1ad841824)

dixonjoel commented:
Are there any examples that are already specifying enum type specializations that we should remove them from?

jasonmreding commented:
IsEnumArrayControl:
- Consider passing out the enum control reference for the array element instead of the array reference itself. This will simplify the caller a bit.

Get Enum Type Specializations:
- Minor: There are a couple of unbundle by name primitives where the Annotations property isn't wired which is kind of confusing. It would be better to just remove that element from the unbundle.

Get Annotations for Parameter:
- Notice the coercion dot on the indicator. That is because we basically have two typedefs for the same thing. One of them comes from ConfigurationParameter_AnnotationsEntry.ctl and the other comes from the grpc-labview generated typedef defined by the v2 measurement service interface. Paul tried to remove any direct dependencies to a version specific interface at this level so here's what I think we should do:
  - Rename ConfigurationParameter_AnnotationsEntry.ctl to Parameter_AnnotationsEntry.ctl.
  - Update Output.ctl to use this typedef rather than the one from the v2 measurement interface.
  - While you're at it, rename Output.ctl to OutputParameter.ctl for consistency.

WesleyTangNationalInstruments commented:
> IsEnumArrayControl: Consider passing out the enum control reference for the array element instead of the array reference itself. This will simplify the caller a bit.

@jasonmreding I've opted not to make any changes for this comment. It seems weird to have `IsEnumControl.vi` return two references to the same control and change `IsEnumArrayControl.vi` to return two references to two different controls. Additionally, we are cleaning up the reference to the array element in `IsEnumArrayControl.vi` and if instead we passed it out as an output, we would move that responsibility into the caller to have to close both the enum and the array reference. From a usage perspective, it seems more intuitive to have them behave the same.

Get Enum Type Specialization.vi
![image](https://github.com/ni/measurementlink-labview/assets/103216564/d38d7388-654f-428f-8bd0-494080256441)

Parameter_AnnotationsEntry.ctl
![image](https://github.com/ni/measurementlink-labview/assets/103216564/dc937cc2-b47c-44a9-9200-3d4a09ea80a1)

OutputParameter.ctl
![image](https://github.com/ni/measurementlink-labview/assets/103216564/16bd26d2-7259-4db0-b7f8-0aee70379f7e)

ConfigurationParameter.ctl
![image](https://github.com/ni/measurementlink-labview/assets/103216564/fbbf9c0f-2d23-477a-a37b-d38877f28f80)

`GetMetadataResponse.ctl`, `MeasurementSignature.ctl`, `Get Metadata.vi`, `Get MetaData from Results.vi`, and `Get Annotations for Parameter.vi` have no visible changes, but they were just updated to reference the renamed `OutputParameter.ctl` and `Parameter_AnnotationsEntry.ctl`.
